napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1521 @@
+        new_z_index = min(self.data._z_index) - 1
+        for index in self.selected_shapes:
+            self.data.update_z_index(index, new_z_index)
+        self.refresh()
+
+    def _move(self, coord):
+        """Moves object at given mouse position and set of indices.
+
+        Parameters
+        ----------
+        coord : sequence of two int
+            Position of mouse cursor in image coordinates.
+        """
+        vertex = self._moving_vertex
+        if (self.mode == 'select' or self.mode == 'add_rectangle' or
+                self.mode == 'add_ellipse' or self.mode == 'add_line'):
+            if len(self.selected_shapes) > 0:
+                self._is_moving = True
+                if vertex is None:
+                    # Check where dragging box from to move whole object
+                    if self._drag_start is None:
+                        center = self._selected_box[-1]
+                        self._drag_start = coord - center
+                    center = self._selected_box[-1]
+                    shift = coord - center - self._drag_start
+                    for index in self.selected_shapes:
+                        self.data.shift(index, shift)
+                    self._selected_box = self._selected_box + shift
+                    self.refresh()
+                elif vertex < 8:
+                    # Corner / edge vertex is being dragged so resize object
+                    box = self._selected_box
+                    if self._fixed_vertex is None:
+                        self._fixed_index = np.mod(vertex+4, 8)
+                        self._fixed_vertex = box[self._fixed_index]
+
+                    size = (box[np.mod(self._fixed_index+4, 8)] -
+                            box[self._fixed_index])
+                    offset = box[-1] - box[-2]
+                    offset = offset/np.linalg.norm(offset)
+                    offset_perp = np.array([offset[1], -offset[0]])
+
+                    if np.mod(self._fixed_index, 2) == 0:
+                        # corner selected
+                        fixed = self._fixed_vertex
+                        new = copy(coord)
+                        if self._fixed_aspect:
+                            if (new - fixed)[0] == 0:
+                                ratio = 1
+                            else:
+                                ratio = abs((new - fixed)[1]/(new - fixed)[0])
+                            if ratio > self._aspect_ratio:
+                                r = self._aspect_ratio/ratio
+                                new[1] = fixed[1]+(new[1]-fixed[1])*r
+                            else:
+                                r = ratio/self._aspect_ratio
+                                new[0] = fixed[0]+(new[0]-fixed[0])*r
+                        dist = np.dot(new-fixed, offset)/np.dot(size, offset)
+                        dist_perp = (np.dot(new-fixed, offset_perp) /
+                                     np.dot(size, offset_perp))
```

Comment:
As I've noted before, this patterns appears multiple times, so it might be worth refactoring — either with a function call or by altering the control flow so that it gets run under more conditions.

Reply:
improved control flow to remove duplicate code